File: 22-generate-parentheses/22-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        string s;
        dfs(res, s, n, 0, 0);
        return res;
    }
    void dfs(vector<string>& res, string s, int n, int left, int right) {
        if (s.size() == 2 * n) {
            res.push_back(s);
            return;
        }
        if (left < n) dfs(res, s + '(', n, left + 1, right);
        if (right < left) dfs(res, s + ')', n, left, right + 1);
    }
};

File: 1-100/22-generate-parentheses/22-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        if(n<1) return new LinkedList<>();
        return generateParenthesis("", n, n);
    }
    
    private List<String> generateParenthesis(String s, int l, int r) {
        List<String> list = new LinkedList<>();
        if(l==0 && r==0) {
            list.add(s);
            return list;
        }
        
        List<String> left = null;
        List<String> right = null;
        if(l>0) left = generateParenthesis(s+"(", l-1, r);
        if(l<r) right = generateParenthesis(s+")", l, r-1);
        
        if(left!=null) list.addAll(left);
        if(right!=null) list.addAll(right);
        return list;
    }
}